Add support for settings aliases/Rename `replication_alter_partitions_sync` to `alter_sync`

antonio2368 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Rename setting `replication_alter_partitions_sync` to `alter_sync`.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

antonio2368 committed:
Add support for settings alias

antonio2368 committed:
Rename setting in test

antonio2368 committed:
Extend support for aliases to constraints and profiles

antonio2368 commented on src/Core/BaseSettings.h:
```diff
@@ -178,6 +178,8 @@ class BaseSettings : public TTraits::Data
     const SettingFieldCustom * tryGetCustomSetting(std::string_view name) const;
 
     std::conditional_t<Traits::allow_custom_settings, CustomSettingMap, boost::blank> custom_settings_map;
+
+    std::map<std::string_view, std::string_view> settings_aliases;
```
```suggestion
```

robot-clickhouse committed:
Automatic style fix

antonio2368 committed:
Remove unused field

alesapin reviewed (commented):
🚀

qoega commented:
Should we consider it backward incompatible change? Or how users can update if they have this setting set ?

antonio2368 commented:
> Should we consider it backward incompatible change? Or how users can update if they have this setting set ?

I tried making it backward compatible and it seems it works based from the test (in tests I used the new setting name only in one place and left it as is in other places).
Also, I updated tests for settings profiles and settings constraints to use old name in config and verified that it correctly applies to the new setting.
If I missed something please tell me, but I think we should be okay.

antonio2368 commented:
Currently I just ignored adding alias (old_name) to the `system.settings` to discourage people from using it.
But I can add it and maybe include an extra column `alias_for`? 🤔 
Also, maybe some people have `SELECT * FROM system.settings WHERE name='old_name'`

tavplubix commented:
There's another thing we should be careful with:
```
dell9510 :) create view v as select 1 settings replication_alter_partitions_sync=2

CREATE VIEW v AS
SELECT 1
SETTINGS replication_alter_partitions_sync = 2

Query id: a026bd91-a39b-4200-8fe1-507963e01caf

Ok.

0 rows in set. Elapsed: 0.013 sec. 

dell9510 :) show create table v

SHOW CREATE TABLE v

Query id: 9432adac-041c-41d7-becc-1b92bc055b98

┌─statement──────────────────────────────────────────────────────────────┐
│ CREATE VIEW default.v
(
    `1` UInt8
) AS
SELECT 1
SETTINGS alter_sync = 2 │
└────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.007 sec. 

```

It potentially may break the recovery of stale Replicated databases. However, seems like the only way to use a setting from Settings.h in a CREATE query is to specify it in the SELECT query when creating a [materialized] view. It it just does not make any sense to specify `replication_alter_partitions_sync` there. So I guess we can ignore it. But we should be careful with settings aliases in the future

serxa commented on src/Core/Settings.h:
```diff
@@ -896,16 +896,17 @@ class IColumn;
     FORMAT_FACTORY_SETTINGS(M)
 
-DECLARE_SETTINGS_TRAITS_ALLOW_CUSTOM_SETTINGS(SettingsTraits, LIST_OF_SETTINGS)
+#define LIST_OF_ALIASES(M) \
+    M("replication_alter_partitions_sync", "alter_sync") \
+
+DECLARE_SETTINGS_TRAITS_ALLOW_CUSTOM_SETTINGS(SettingsTraits, LIST_OF_SETTINGS, LIST_OF_ALIASES)
```
To avoid creating this additional `LIST_OF_ALIASES` and `NO_ALIASES` macro you can instead do something like this:
```cpp
#define COMMON_SETTINGS(M, ALIAS) \
  M(UInt64, alter_sync, 1, "Text", 0) ALIAS(replication_alter_partitions_sync)
  ```
It would require rework of existing code generation logic, but not too hard I guess. Here is possible implementation draft https://godbolt.org/z/c9bYbozaY. WDYT?

serxa replied:
Great, even more compact! The only thing is that now it should be in "quotes". Not sure what is better. At first we would not have more than single alias per setting. And it is different from original declaration w/o quotes. So, up to you, I like both variants.

serxa commented on src/Access/SettingsConstraints.cpp:
```diff
@@ -36,9 +36,43 @@ void SettingsConstraints::clear()
+            resolved_name);
+    };
+
+    if (auto resolved_name = SettingsTraits::resolveName(name); resolved_name != name)
+    {
+        write_warning_for_changed(resolved_name, name);
+        return resolved_name;
+    }
+
+    if (auto resolved_name = MergeTreeSettingsTraits::resolveName(name); resolved_name != name)
+    {
+        write_warning_for_changed(resolved_name, name);
+        return resolved_name;
+    }
```
This is a little bit wrong. It does not take `MERGE_TREE_SETTINGS_PREFIX` into account. I think it may be better to integrate this logic into `resolveSetting.h`.

antonio2368 committed:
Merge branch 'master' into add-support-for-settings-alias

antonio2368 committed:
Better alias definition

antonio2368 committed:
Merge branch 'master' into add-support-for-settings-alias

antonio2368 committed:
Merge branch 'master' into add-support-for-settings-alias

antonio2368 committed:
Better constraint resolution

antonio2368 committed:
Merge branch 'master' into add-support-for-settings-alias

antonio2368 committed:
Update docs

antonio2368 committed:
Update tests and system.settings

antonio2368 commented:
@tavplubix the VIEW "issue" should not happen anymore, it was a side effect of my previous bad implementation decision.

@serxa I added changes that should correctly cover MergeTreeSettings in constraints and your way of defining aliases because (really cool).
Also I updated system.settings, it now will show aliases and it has new column (`alias_for`).

antonio2368 committed:
Extract common code

antonio2368 committed:
Fix fast test

antonio2368 committed:
Merge branch 'master' into add-support-for-settings-alias

antonio2368 committed:
Correctly use merge tree setting constraints

antonio2368 committed:
Use correct name resolution

serxa commented:
@vitlibar please, also take a look at SettingsConstraints refactoring

antonio2368 committed:
Merge branch 'master' into add-support-for-settings-alias

antonio2368 committed:
Merge branch 'master' into add-support-for-settings-alias
